### PR TITLE
fix: adds {shadowOptions: null} support to fast-ssr

### LINKED ISSUE #6394: fix: support `shadowOptions: null` config in fast-ssr
# 🐛 Bug Report
fast-element supports rendering a component's template to the light DOM by assigning [`shadowOptions: null` when defining the component.](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-element/docs/guide/working-with-shadow-dom.md#shadow-dom-configuration).

There is currently no handling of this in `@microsoft/fast-ssr`, meaning the component will render it's template to declarative shadow DOM.

## 🤔 Expected Behavior
A component defined with null shadowOptions should emit its template as child-elements with no declarative shadow dom. Styles should be emitted as the first children of the element.

### PR DESCRIPTION
**Pull request by nicholasrice**
# Pull Request

## 📖 Description
This PR adds support for elements declared with `{shadowOptions: null}` to the SSR renderer. Components declared this way will not have declarative shadow-dom emitted, with templates being rendered to light DOM and styles rendered as the first children of the element.

This PR also addresses a `shadowRoot` leak from `ElementController` via `ElementStyles` and `StyleStrategy`. This PR changes the behavior of `ElementController.addStyles()` by invoking `ElementStyles.addStylesTo()` with the custom element host instead of the element's shadowRoot. Resolution of the `shadowRoot` was moved to the internals of `StyleStrategy` itself, preventing the leak.

### 🎫 Issues
fixes #6394

### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/web-components/fast-ssr/src/template-renderer/template-renderer.spec.ts [hunk 1]
```diff
@@ -124,7 +125,51 @@ test.describe("TemplateRenderer", () => {
 
             expect(consolidate(result)).toBe(`<with-host-attributes  id="foo" static="static" dynamic="dynamic" bool-true><template shadowroot=\"open\">value<slot></slot></template></with-host-attributes>`);
         });
-    })
+    });
+    test.describe("rendering a FAST element configured with a null shadowOptions config", () => {
```
Comment (EisenbergEffect) [inline packages/web-components/fast-ssr/src/template-renderer/template-renderer.spec.ts:129]: Might be good to have a test that has `shadowOptions: null` and also has a template.

File: packages/web-components/fast-ssr/src/configure-fast-element.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+function removeStyles(this: ElementController, styles: ElementStyles) {
+    try {
+        return originalRemoveStyles.call(this, styles);
+    } catch (e) {
+        styles.strategy.removeStylesFrom(this.source);
+        styles.behaviors?.forEach(b => this.removeBehavior(b));
+    }
+}
+
+ElementController.prototype.addStyles = addStyles;
```
Comment (EisenbergEffect) [inline packages/web-components/fast-ssr/src/configure-fast-element.ts:65]: This makes me a little uncomfortable...

File: packages/web-components/fast-ssr/src/template-renderer/template-renderer.spec.ts (hunk 1 above)
Comment (nicholasrice) [reply packages/web-components/fast-ssr/src/template-renderer/template-renderer.spec.ts:129]: oops I thought I added that, good catch

**Comment by nicholasrice** [comment]
Going to re-approach this

File: packages/web-components/fast-ssr/src/configure-fast-element.ts (hunk 2 above)
Comment (nicholasrice) [reply packages/web-components/fast-ssr/src/configure-fast-element.ts:65]: Me too - fixed =D

File: packages/web-components/fast-element/src/components/element-controller.ts [hunk 3]
```diff
@@ -260,14 +266,17 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
         }
 
         const source = this.source;
-        const target: StyleTarget =
-            getShadowRoot(source) ?? ((source.getRootNode() as any) as StyleTarget);
 
         if (styles instanceof HTMLElement) {
+            // This won't be supported in SSR anyway, so we can just handle client-side
+            // Get shadowRoot
+            // if shadowRoot append the element to the shadowRoot
```
Comment (nicholasrice) [inline packages/web-components/fast-element/src/components/element-controller.ts:273]: note-to-self: remove these notes, they're not useful here.

**Comment by EisenbergEffect** [comment]
I think we'll need to talk about this more. There are some additional things that come into play if we have to move the StyleElementStyles into the core. We'll need to fix up the kernel stuff and potentially move the polyfill back into the core as well since there will be a lot less there. @nicholasrice ping me when you are free tomorrow and we can discuss.

**Review by EisenbergEffect — APPROVED** [review]

File: packages/web-components/fast-element/src/components/element-controller.ts [hunk 4]
```diff
@@ -470,3 +468,117 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
...
+        this.styleClass = nextStyleId();
+    }
+
+    public addStylesTo(target: StyleTarget): void {
+        target = usableStyleTarget(normalizeStyleTarget(target));
+
+        const styles = this.styles;
+        const styleClass = this.styleClass;
+
+        for (let i = 0; i < styles.length; i++) {
```
Comment (janechu) [inline packages/web-components/fast-element/src/components/element-controller.ts:558]:
```suggestion
        for (let i = 0, stylesLength = styles.length; i < stylesLength; i++) {
```

**Review by janechu — APPROVED** [review]

File: packages/web-components/fast-element/src/components/element-controller.ts (hunk 4 above)
Comment (nicholasrice) [reply packages/web-components/fast-element/src/components/element-controller.ts:558]: Going to leave this as-is mostly so I don't have to rebuild.
I just moved this code from another file, so it was pre-existing.

**Pull request merged**